### PR TITLE
FOGL-7001 Add basic healthcheck option

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Signed-off-by: Mark Riddoch <mark@dianomic.com>

### CODE REVIEW

**Review by pintomax — APPROVED** [review]

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
1. Unresponsive services are not showing, due to wrong variable

```
$ ./scripts/fledge healthcheck
Fledge Healthcheck
==================

Some services have failed at least once today
Failures Service
      1 "bench"

Service State
The services %s are unresponsive
The services "bench" have failed
The services "sine" "omf" are shutdown

Healthcheck completed
```

2. Two services are in failed state.

```
~/fledge(FOGL-7001) $ ./scripts/fledge healthcheck
Fledge Healthcheck
==================

Some services have failed at least once today
Failures Service
      1 "random" "bench"

Service State
The services "bench" "random" have failed
The services "sine" "omf" are shutdown

Healthcheck completed

```
Failures Service
      1 "random" 
      2 "bench"

Or remove numbering and do with comma separated or space separated

Same applies for Service State section.

3. Also when authentication is required/mandatory. I am not able to test it as it throws failed authentication to stop fledge. (Entered password by default)
```
$ ./scripts/fledge status
Fledge v2.0.0 running.
Fledge Uptime:  261 seconds.
Fledge records: 200 read, 0 sent, 0 purged.
Fledge requires authentication.
=== Fledge services:
fledge.services.core
fledge.services.storage --address=0.0.0.0 --port=38025
=== Fledge tasks:

~/fledge(FOGL-7001) $ curl -sX GET localhost:8081/fledge/ping
{"uptime": 279, "dataRead": 200, "dataSent": 0, "dataPurged": 0, "authenticationOptional": false, "serviceName": "Fledge", "hostName": "aj-ThinkPad-E450", "ipAddresses": ["192.168.1.8"], "health": "green", "safeMode": false, "version": "2.0.0"}aj@aj-ThinkPad-E450 ~/fledge(FOGL-7001) $ 

~/fledge(FOGL-7001) $ ./scripts/fledge healthcheck
Fledge Healthcheck
==================

Password: Failed authentication when attempting to stop fledge.
Authentication failed.

```

**Comment by MarkRiddoch** [comment]
> 1. Unresponsive services are not showing, due to wrong variable
> 
> ```
> $ ./scripts/fledge healthcheck
> Fledge Healthcheck
> ==================
> 
> Some services have failed at least once today
> Failures Service
>       1 "bench"
> 
> Service State
> The services %s are unresponsive
> The services "bench" have failed
> The services "sine" "omf" are shutdown
> 
> Healthcheck completed
> ```
> 
>  2. Two services are in failed state.
> 
> ```
> ~/fledge(FOGL-7001) $ ./scripts/fledge healthcheck
> Fledge Healthcheck
> ==================
> 
> Some services have failed at least once today
> Failures Service
>       1 "random" "bench"
> 
> Service State
> The services "bench" "random" have failed
> The services "sine" "omf" are shutdown
> 
> Healthcheck completed
> ```
> 
> Failures Service 1 "random" 2 "bench"
> 
> Or remove numbering and do with comma separated or space separated

It's not numbering it is the count of the failures. They have each failed once
> 
> Same applies for Service State section.
> 
> 3. Also when authentication is required/mandatory. I am not able to test it as it throws failed authentication to stop fledge. (Entered password by default)
> 
> ```
> $ ./scripts/fledge status
> Fledge v2.0.0 running.
> Fledge Uptime:  261 seconds.
> Fledge records: 200 read, 0 sent, 0 purged.
> Fledge requires authentication.
> === Fledge services:
> fledge.services.core
> fledge.services.storage --address=0.0.0.0 --port=38025
> === Fledge tasks:
> 
> ~/fledge(FOGL-7001) $ curl -sX GET localhost:8081/fledge/ping
> {"uptime": 279, "dataRead": 200, "dataSent": 0, "dataPurged": 0, "authenticationOptional": false, "serviceName": "Fledge", "hostName": "aj-ThinkPad-E450", "ipAddresses": ["192.168.1.8"], "health": "green", "safeMode": false, "version": "2.0.0"}aj@aj-ThinkPad-E450 ~/fledge(FOGL-7001) $ 
> 
> ~/fledge(FOGL-7001) $ ./scripts/fledge healthcheck
> Fledge Healthcheck
> ==================
> 
> Password: Failed authentication when attempting to stop fledge.
> Authentication failed.
> ```

File: scripts/fledge [hunk 1]
```diff
@@ -636,6 +637,116 @@ fledge_purge() {
...
+	else
+		result=`curl -s ${REST_API_URL}/fledge/health/storage`
+	fi
+	a=`echo $result |jq 'select (.disk.usage > 80) .disk.usage'`
+	if [[ "$a" != "" ]]; then
+		echo The disk space in the storage directory is low, $a%
+		echo
+	fi
+
+	service rsyslog status >/dev/null
```
Comment (ashish-jabble) [inline scripts/fledge:698]: Assuming that rsyslog is available for all Fledge supported platforms

File: scripts/fledge [hunk 2]
```diff
@@ -636,6 +637,116 @@ fledge_purge() {
...
+		if [[ "${token}" =~ "failed" ]]; then
+			fledge_log "info" "Failed authentication when attempting to stop fledge." "all" "pretty"
+			echo "Authentication failed."
+			exit 0
+		fi
+		result=`curl -k -s -H "authorization: $token" ${REST_API_URL}/fledge/health/logging`
+		auth="required"
+	fi
+
+	a=`echo $result | jq '.levels[] | {name,level}' | jq 'select (.level == "info") .name'`
```
Comment (ashish-jabble) [inline scripts/fledge:667]: Can we avoid `jq` here? Otherwise we need to add this APT package in Debian Packaging.

File: scripts/fledge (hunk 1 above)
Comment (MarkRiddoch) [reply scripts/fledge]: Without it we get no log output. This was an issue at BRP, hence the inclusion

File: scripts/fledge (hunk 2 above)
Comment (MarkRiddoch) [reply scripts/fledge:667]: It's a lot easier with it. I thought we already had it. If not we could do it with some horrid python scripting!

**Review by pintomax — DISMISSED** [review]

File: scripts/fledge [hunk 3]
```diff
@@ -734,23 +737,26 @@ fledge_healthcheck() {
 	else
 		result=`curl -s ${REST_API_URL}/fledge/service`
 	fi
-	a=`echo $result | jq '.services[] | select (.status == "unresponsive") .name'`
+#	a=`echo $result | jq '.services[] | select (.status == "unresponsive") .name'`
```
Comment (pintomax) [inline scripts/fledge:740]: remove commented line

File: scripts/fledge [hunk 4]
```diff
@@ -734,23 +737,26 @@ fledge_healthcheck() {
...
+	a=`echo $result | python3 -m scripts.common.service_status unresponsive`
 	if [[ "$a" != "" ]]; then
 		echo The following services are unresponsive:
 	        for i in $a; do
 			echo $i | sed -e s/\"//g -e 's/^/    /'
 		done
 		issue="some"
 	fi
-	a=`echo $result | jq '.services[] | select (.status == "failed") .name'`
+	#a=`echo $result | jq '.services[] | select (.status == "failed") .name'`
```
Comment (pintomax) [inline scripts/fledge:749]: remove commented line

File: scripts/fledge [hunk 5]
```diff
@@ -734,23 +737,26 @@ fledge_healthcheck() {
...
+	a=`echo $result | python3 -m scripts.common.service_status failed`
 	if [[ "$a" != "" ]]; then
 		echo The following services have failed:
 	        for i in $a; do
 			echo $i | sed -e s/\"//g -e 's/^/    /'
 		done
 		issue="some"
 	fi
-	a=`echo $result | jq '.services[] | select (.status == "shutdown") .name'`
+#	a=`echo $result | jq '.services[] | select (.status == "shutdown") .name'`
```
Comment (pintomax) [inline scripts/fledge:758]: remove commented line

**Review by pintomax — CHANGES_REQUESTED** [review]

File: scripts/common/audittime.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import json
+import sys
+
+if __name__ == '__main__':
+   json = json.loads(sys.stdin.readline())
+   ts = sys.argv[1]
+   for audit in json["audit"]:
+       if audit["timestamp"].startswith(ts):
+           print(audit["details"]["name"])
```
Comment (ashish-jabble) [inline scripts/common/audittime.py:17]: Bad indentation; indentation is not a multiple of 4 spaces.

File: scripts/common/disk_usage.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+"""
+import json
+import sys
+
+if __name__ == '__main__':
+   json = json.loads(sys.stdin.readline())
+   item = json["disk"]
+   usage=item["usage"]
+   if usage >= int(sys.argv[1]):
+       print(usage)
```
Comment (ashish-jabble) [inline scripts/common/disk_usage.py:15]: Bad indentation; indentation is not a multiple of 4 spaces.

Also missing whitespace around operator - `usage = `

File: scripts/common/service_status.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import json
+import sys
+
+if __name__ == '__main__':
+   json = json.loads(sys.stdin.readline())
+   state = sys.argv[1]
+   for service in json["services"]:
+       if service["status"] == state:
+           print(service["name"])
```
Comment (ashish-jabble) [inline scripts/common/service_status.py:14]: Bad indentation; indentation is not a multiple of 4 spaces.

File: scripts/common/service_status.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+"""
+Utility to extract the names of services in a particular state giventhe output of
```
Comment (ashish-jabble) [inline scripts/common/service_status.py:2]: space b/w - given the

**Comment by ashish-jabble** [comment]
@MarkRiddoch You may have missed given below issue in my above comments and still not resolved.
> Also when authentication is required/mandatory. I am not able to test it as it throws failed authentication to stop fledge. (Entered password by default)
$ ./scripts/fledge status
Fledge v2.0.0 running.
Fledge Uptime:  261 seconds.
Fledge records: 200 read, 0 sent, 0 purged.
Fledge requires authentication.
=== Fledge services:
fledge.services.core
fledge.services.storage --address=0.0.0.0 --port=38025
=== Fledge tasks:
~/fledge(FOGL-7001) $ curl -sX GET localhost:8081/fledge/ping
{"uptime": 279, "dataRead": 200, "dataSent": 0, "dataPurged": 0, "authenticationOptional": false, "serviceName": "Fledge", "hostName": "aj-ThinkPad-E450", "ipAddresses": ["192.168.1.8"], "health": "green", "safeMode": false, "version": "2.0.0"}aj@aj-ThinkPad-E450 ~/fledge(FOGL-7001) $ 
~/fledge(FOGL-7001) $ ./scripts/fledge healthcheck
Fledge Healthcheck
==================
Password: Failed authentication when attempting to stop fledge.
Authentication failed.

File: scripts/common/audittime.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+import json
+import sys
+
+if __name__ == '__main__':
+    json = json.loads(sys.stdin.readline())
+    ts = sys.argv[1]
+    for audit in json["audit"]:
+         if audit["timestamp"].startswith(ts):
+              print(audit["details"]["name"])
```
Comment (ashish-jabble) [inline scripts/common/audittime.py:17]: Now over indented;  indentation is not a multiple of 4

File: scripts/common/disk_usage.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+The script takes as input the output of the GET /fledge/health/logging
+or GET /fledge/health/storage API call
+"""
+import json
+import sys
+
+if __name__ == '__main__':
+    json = json.loads(sys.stdin.readline())
+    item = json["disk"]
+    usage=item["usage"]
```
Comment (ashish-jabble) [inline scripts/common/disk_usage.py:13]: Add whitespace around operator; `usage = `

File: scripts/common/disk_usage.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+    if usage >= int(sys.argv[1]):
+         print(usage)
```
Comment (ashish-jabble) [inline scripts/common/disk_usage.py:15]: Now over indented;  indentation is not a multiple of 4

File: scripts/common/service_status.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+import json
+import sys
+
+if __name__ == '__main__':
+    json = json.loads(sys.stdin.readline())
+    state = sys.argv[1]
+    for service in json["services"]:
+         if service["status"] == state:
+              print(service["name"])
```
Comment (ashish-jabble) [inline scripts/common/service_status.py:14]: Now over indented;  indentation is not a multiple of 4

**Review by ashish-jabble — APPROVED** [review]
Now, authentication mandatory scenario looks good. 
After unsetting USERNAME environment variable in my dev environment.

**Review by pintomax — APPROVED** [review]

**Pull request merged**